pkg/devfile/adapters/kubernetes/utils/utils.go

```diff
@@ -17,9 +17,9 @@ import (
 )
 
 // ComponentExists checks whether a deployment by the given name exists
-func ComponentExists(client kclient.Client, name string) bool {
-	_, err := client.GetDeploymentByName(name)
-	return err == nil
+func ComponentExists(client kclient.Client, name string) (bool, error) {
+	deployment, err := client.GetDeploymentByName(name)
+	return len(deployment.GetName()) != 0, err
```

Comment:
if deployment is not found, won't it be nil?

Reply:
This is a weird behaviour from kube client api. It ALWAYS returns a deployment object even if there is no deployment. For example when the deployment is not found, kube api returns 
```
&Deployment{ObjectMeta:{      0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] []  []},Spec:DeploymentSpec{Replicas:nil,Selector:nil,Template:{{      0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] []  []} {[] [] [] []  <nil> <nil>  map[]   <nil>  false false false <nil> nil []   nil  [] []  <nil> nil [] <nil> <nil> <nil> map[] []}},Strategy:DeploymentStrategy{Type:,RollingUpdate:nil,},MinReadySeconds:0,RevisionHistoryLimit:nil,Paused:false,ProgressDeadlineSeconds:nil,},Status:DeploymentStatus{ObservedGeneration:0,Replicas:0,UpdatedReplicas:0,AvailableReplicas:0,UnavailableReplicas:0,Conditions:[]DeploymentCondition{},ReadyReplicas:0,CollisionCount:nil,},} err: <nil>
```

I cant check for `nil` or `reflect.DeepEqual` an empty deployment struct. I found that when the deployment is not there, the name length is always 0. 

Any other suggestions welcome! But I think alternative would be to replace `GetDeploymentByName` logic with the `List` API like I suggested earlier here https://github.com/openshift/odo/pull/3405#discussion_r446437850 and try getting the deployment with labels.

Reply:
Maybe let's move the error checking logic from GetDeploymentByName() to here and leave that unchanged. I feels that GetDeploymentByName throws the error so other functions can modify the error checking according to their needs.
```
	if kerrors.IsNotFound(err) {
		klog.V(4).Infof("Deployment %s not found", name)
		return false, nil
	} else if err!=nil {
		return true, err
	}
	return deployment!=nil, err
```

Reply:
true makes much more sense, updated